[VPlan] Introduce constant folder in simplifyRecipe

artagnon opened the pull request:
Introduce a VPlan-level constant folder in simplifyRecipe that tries to fold a recipe to a constant using TargetFolder.

artagnon commented:
Okay, this is now fixed to get the Plan on-demand.

artagnon commented:
@nikic I'm not sure how the ConstantExpr creation ties into your effort to remove constant-expressions. Could you kindly have a look?

nikic commented:
> @nikic I'm not sure how the ConstantExpr creation ties into your effort to remove constant-expressions. Could you kindly have a look?

Please only use APIs defined in https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/Analysis/ConstantFolding.h.

I'd consider only extracting the IR values and delegating the actual folding to https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/Analysis/TargetFolder.h.

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorizationPlanner.h:
```diff
@@ -66,6 +67,12 @@ class VPBuilder {
+  VPValue *getOrAddLiveIn(Value *V) {
+    if (BB)
+      return BB->getPlan()->getOrAddLiveIn(V);
+    return new VPValue(V);
```
This will introduce a leak?

artagnon replied:
Will it? I thought the Builder is used in two different ways: when the insertion point is set, and when the insertion point isn't set. For the case when the insertion point isn't set, wouldn't the following introduce a leak, if this introduces a leak?

```cpp
template <typename T> T *tryInsertInstruction(T *R) {
    if (BB)
      BB->insert(R, InsertPt);
    return R;
  }
  VPInstruction *createInstruction(unsigned Opcode,
                                   ArrayRef<VPValue *> Operands, DebugLoc DL,
                                   const Twine &Name = "") {
    return tryInsertInstruction(new VPInstruction(Opcode, Operands, DL, Name));
  }
```

fhahn commented on llvm/test/Transforms/LoopVectorize/X86/replicate-recipe-with-only-first-lane-used.ll:
```diff
@@ -12,56 +12,10 @@ define void @replicate_udiv_with_only_first_lane_used(i32 %x, ptr %dst, i64 %d)
+; CHECK-NEXT:    [[TMP11:%.*]] = getelementptr i16, ptr [[DST]], i64 0
 ; CHECK-NEXT:    br label %[[VECTOR_BODY:.*]]
 ; CHECK:       [[VECTOR_BODY]]:
-; CHECK-NEXT:    [[INDEX:%.*]] = phi i64 [ 0, %[[VECTOR_PH]] ], [ [[INDEX_NEXT:%.*]], %[[PRED_UDIV_CONTINUE14:.*]] ]
```
We should probably add a variant of the test where it cannot be constant folded, to preserve the original test coverage.

artagnon replied:
Sounds good: will do.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanConstantFolder.h:
```diff
@@ -0,0 +1,91 @@
+//
+//===----------------------------------------------------------------------===//
+
+#include "VPlanValue.h"
```
Should have include guards?

artagnon replied:
Sorry about this oversight: will fix.

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorizationPlanner.h:
```diff
@@ -25,7 +25,7 @@
 #define LLVM_TRANSFORMS_VECTORIZE_LOOPVECTORIZATIONPLANNER_H
 
 #include "VPlan.h"
-#include "llvm/ADT/SmallSet.h"
```
Is this intentional?

artagnon replied:
Just thought I'd clean up unused includes when making a change to the file: can revert if undesired?

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanConstantFolder.h:
```diff
@@ -0,0 +1,91 @@
+  ConstantFolder Folder;
+
+  Constant *getIRConstant(VPValue *V) const {
+    return dyn_cast_if_present<Constant>(V->getUnderlyingValue());
```
Using the underlying IR value is only safe for live-in VPValues; for recipes, the underlying value should not have any semantic meaning (not 100% true yet for all recipes, but we are getting there)

artagnon replied:
Oh, I saw that this wasn't true in practice, and even put up https://github.com/llvm/llvm-project/pull/128886, which includes some improvements. I do have a question though: isn't the defining recipe created with several intermediate Builder-create calls? Why should the defining recipe be present when we're in the process of building recipes?

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -533,8 +533,8 @@ createScalarIVSteps(VPlan &Plan, InductionDescriptor::InductionKind Kind,
   VPCanonicalIVPHIRecipe *CanonicalIV = Plan.getCanonicalIV();
-  VPSingleDefRecipe *BaseIV = Builder.createDerivedIV(
-      Kind, FPBinOp, StartV, CanonicalIV, Step, "offset.idx");
+  VPValue *BaseIV = Builder.createDerivedIV(Kind, FPBinOp, StartV, CanonicalIV,
```
Is this intentional?

artagnon replied:
Yes, this is due to the change in the return type of createDerivedIV, nothing more.

fhahn replied:
But the return type did not change AFAICT?

artagnon replied:
See several lines below:

```cpp
    BaseIV = Builder.createScalarCast(Instruction::Trunc, BaseIV, TruncTy, DL);
```

fhahn reviewed (commented):
Hmm, I am a bit surprised there's not more to fold in the existing test cases. Any chance we could at least one  test covering each fold?

artagnon reviewed (commented):
> Hmm, I am a bit surprised there's not more to fold in the existing test cases. Any chance we could at least one test covering each fold?

I was actually very surprised when I authored the patch and saw so few test changes. Yes, will write tests to cover each fold, along with the other pending items in this review.

artagnon committed:
[LV] Pre-commit tests for const-folder

artagnon commented:
Gentle ping.

nikic commented on llvm/lib/Transforms/Vectorize/VPlanConstantFolder.h:
```diff
@@ -0,0 +1,98 @@
+namespace llvm {
+class VPConstantFolder {
+private:
+  ConstantFolder Folder;
```
It is better to use TargetFolder instead of ConstantFolder.

artagnon replied:
I know, but the DataLayout is not readily available at the VPlan level, and I wonder if it makes sense to thread it into VPlan?

nikic replied:
If you want to do constant folding, then yes, you should thread it into VPlan.

artagnon replied:
Thanks! There are lots of test changes now, and I wonder if the patch should be split out.

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorizationPlanner.h:
```diff
@@ -180,31 +186,45 @@ class VPBuilder {
 
   VPValue *createAnd(VPValue *LHS, VPValue *RHS, DebugLoc DL = {},
                      const Twine &Name = "") {
+    if (auto *V = Folder.foldAnd(LHS, RHS))
```
I am not sure if it is possible, but it would be nice if the folder could just have a single interface that takes opcode + operands and then we can have a single place like `createInstruction` that queries the folder, instead of having to update each create* function here and add a corresponding `foldXXX` function to the folder, which seems quite verbose.

artagnon replied:
I thought about this, but I fear we'll have to match the VPInstruction using a switch, and dispatch appropriately: not sure if this is preferable, so let's keep this thread open.

artagnon commented:
Gentle ping. The only pending item for discussion is whether the code can somehow be made more compact.

fhahn commented on llvm/test/Transforms/LoopVectorize/AArch64/intrinsiccost.ll:
```diff
@@ -35,13 +35,13 @@ define void @saddsat(ptr nocapture readonly %pSrc, i16 signext %offset, ptr noca
 ; CHECK-NEXT:    [[NEXT_GEP3:%.*]] = getelementptr i8, ptr [[PDST:%.*]], i64 [[OFFSET_IDX2]]
-; CHECK-NEXT:    [[TMP1:%.*]] = getelementptr i8, ptr [[NEXT_GEP]], i64 16
-; CHECK-NEXT:    [[WIDE_LOAD:%.*]] = load <8 x i16>, ptr [[NEXT_GEP]], align 2
+; CHECK-NEXT:    [[TMP1:%.*]] = getelementptr i8, ptr [[NEXT_GEP3]], i64 16
```
Interesting, it looks like most changes are just slight re-ordering?

lukel97 commented on llvm/test/Transforms/LoopVectorize/induction-unroll-novec.ll:
```diff
@@ -1,29 +1,59 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 5
```
Precommit this regeneration?

artagnon commented:
Thanks for the review: I've now moved the logic to simplifyRecipe, leaving a couple of TODOs. If this is the desired approach, we can improve upon it in follow-ups.

artagnon commented on llvm/lib/Transforms/Vectorize/VPlanPatternMatch.h:
```diff
@@ -173,8 +173,8 @@ struct Recipe_match {
 
-    assert(R->getNumOperands() == std::tuple_size<Ops_t>::value &&
-           "recipe with matched opcode the expected number of operands");
+    if (R->getNumOperands() != std::tuple_size<Ops_t>::value)
+      return false;
```
This change is due to m_GetElementPtr asserting. It's actually m_PtrAdd, but we can take care of it in follow-ups.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1076,15 +1078,39 @@ static void simplifyRecipe(VPRecipeBase &R, VPTypeAnalysis &TypeInfo) {
+
+  // Constant folding. TODO: cast and cmp.
+  VPConstantFolder Folder(DL);
+  VPlan *Plan = R.getParent()->getPlan();
+  if (match(&R, m_BinaryAnd(m_VPValue(X), m_VPValue(Y))))
+    if (Value *V = Folder.foldAnd(X, Y))
+      R.getVPSingleValue()->replaceAllUsesWith(Plan->getOrAddLiveIn(V));
+  if (match(&R, m_BinaryOr(m_VPValue(X), m_VPValue(Y))))
+    if (Value *V = Folder.foldOr(X, Y))
+      R.getVPSingleValue()->replaceAllUsesWith(Plan->getOrAddLiveIn(V));
+  if (match(&R, m_Not(m_VPValue(X))))
+    if (Value *V = Folder.foldNot(X))
+      R.getVPSingleValue()->replaceAllUsesWith(Plan->getOrAddLiveIn(V));
+  if (match(&R, m_LogicalAnd(m_VPValue(X), m_VPValue(Y))))
+    if (Value *V = Folder.foldLogicalAnd(X, Y))
+      R.getVPSingleValue()->replaceAllUsesWith(Plan->getOrAddLiveIn(V));
+  if (match(&R, m_Select(m_VPValue(X), m_VPValue(Y), m_VPValue(Z))))
+    if (Value *V = Folder.foldSelect(X, Y, Z))
+      R.getVPSingleValue()->replaceAllUsesWith(Plan->getOrAddLiveIn(V));
+  if (match(&R, m_GetElementPtr(m_VPValue(X), m_VPValue(Y))))
+    if (Value *V = Folder.foldPtrAdd(
+            X, Y, cast<VPRecipeWithIRFlags>(R).getGEPNoWrapFlags()))
+      R.getVPSingleValue()->replaceAllUsesWith(Plan->getOrAddLiveIn(V));
```
Would it be possible to have a single `tryToConstantFold(unsigned Opcode, ArrayRef<VPValue *> Ops)` that calls the corresponding `Folder.foldXXX` functions? Then there would only be a single place where we need to add folding support for new opcodes?

The relevant recipe classes should provide a `::getOpcode()` helper, which is used by the constant folder. Maybe it would be possible to use that?

artagnon replied:
Just to clarify: what should I call tryToConstantFold with? `R->getOpcode()`, and is there some API for getting the ops in a generic fashion? I wasn't able to find this, which is why I used the recipe matchers.

fhahn replied:
I think it would need something like a type switch to cast to specific recipe types that have `getOpcode` and operands can be accessed by `operands()`.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanConstantFolder.h:
```diff
@@ -0,0 +1,100 @@
+
+namespace llvm {
+class VPConstantFolder {
+private:
```
```suggestion
```

should not be needed as `class` defaults to private?

fhahn commented on llvm/test/Transforms/LoopVectorize/constantfolder.ll:
```diff
@@ -0,0 +1,360 @@
+  ret void
+}
+;.
+; CHECK: [[LOOP0]] = distinct !{[[LOOP0]], [[META1:![0-9]+]], [[META2:![0-9]+]]}
+; CHECK: [[META1]] = !{!"llvm.loop.isvectorized", i32 1}
+; CHECK: [[META2]] = !{!"llvm.loop.unroll.runtime.disable"}
+; CHECK: [[LOOP3]] = distinct !{[[LOOP3]], [[META2]], [[META1]]}
+; CHECK: [[LOOP4]] = distinct !{[[LOOP4]], [[META1]], [[META2]]}
+; CHECK: [[LOOP5]] = distinct !{[[LOOP5]], [[META2]], [[META1]]}
+; CHECK: [[LOOP6]] = distinct !{[[LOOP6]], [[META1]], [[META2]]}
+; CHECK: [[LOOP7]] = distinct !{[[LOOP7]], [[META2]], [[META1]]}
+; CHECK: [[LOOP8]] = distinct !{[[LOOP8]], [[META1]], [[META2]]}
+; CHECK: [[LOOP9]] = distinct !{[[LOOP9]], [[META2]], [[META1]]}
+; CHECK: [[LOOP10]] = distinct !{[[LOOP10]], [[META1]], [[META2]]}
+; CHECK: [[LOOP11]] = distinct !{[[LOOP11]], [[META2]], [[META1]]}
+; CHECK: [[LOOP12]] = distinct !{[[LOOP12]], [[META1]], [[META2]]}
+; CHECK: [[LOOP13]] = distinct !{[[LOOP13]], [[META2]], [[META1]]}
+;.
```
might be good to generate with `--check-globals=none`

fhahn commented on llvm/test/Transforms/LoopVectorize/constantfolder.ll:
```diff
@@ -0,0 +1,360 @@
+  %or = or i64 %d, %const.1
+  store i64 %or, ptr %dst, align 8
+  %iv.next = add i64 %iv, 1
+  %cmp = icmp ult i64 %iv.next, 101
```
might be worth using `100` instead, so the scalar loop doesn't need to execute?

artagnon commented:
Thanks for the review. The new design is looking good.

artagnon commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -4460,6 +4460,7 @@ static bool willGenerateVectors(VPlan &Plan, ElementCount VF,
       case VPDef::VPWidenSelectSC:
       case VPDef::VPBlendSC:
       case VPDef::VPFirstOrderRecurrencePHISC:
+      case VPDef::VPHistogramSC:
```
This is an underlying bug. Split out into #136712.

fhahn reviewed (commented):
I suspect this will fix https://github.com/llvm/llvm-project/issues/125278 by folding away the redundant casts?

lukel97 commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -931,9 +932,21 @@ static void recursivelyDeleteDeadRecipes(VPValue *V) {
+            VPlan *Plan = R.getParent()->getPlan();
+            ArrayRef<VPValue *> Ops(I->op_begin(), I->op_end());
+            if (Value *V = Folder.tryToConstantFold(R, I->getOpcode(), Ops))
+              R.getVPSingleValue()->replaceAllUsesWith(Plan->getOrAddLiveIn(V));
```
We should probably return from simplifyRecipe here if R is now dead?

artagnon commented:
Gentle ping.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -937,10 +938,84 @@ static void recursivelyDeleteDeadRecipes(VPValue *V) {
   using namespace llvm::VPlanPatternMatch;
 
+  // Constant folding.
+  VPConstantFolder Folder(DL, TypeInfo);
```
now that everything is in a simple compact `tryToConstantFold`, is the `VPConstantFolder` still needed or can we simply pass `TypeInfo` and `DL` to the function, as the Folder instance won't get removed?

fhahn reviewed (commented):
Thanks for the latest updates.

It looks like some checks related to pointers may be missing, I am seeing a crash with the following IR

```
; opt -p loop-vectorize

target datalayout = "e-m:o-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-n32:64-S128-Fn32"
target triple = "arm64-apple-macosx15.0.0"

@postscale = external constant [64 x float]

define void @test(ptr %data) {
entry:
  br label %for.body

for.body:                                         ; preds = %for.body, %entry
  %indvars.iv78 = phi i64 [ %indvars.iv.next79, %for.body ], [ 0, %entry ]
  %0 = or disjoint i64 %indvars.iv78, 1
  %arrayidx142 = getelementptr [64 x float], ptr @postscale, i64 0, i64 %0
  %1 = load float, ptr %arrayidx142, align 4, !tbaa !0
  %2 = tail call i64 @llvm.lrint.i64.f32(float %1)
  %conv145 = trunc i64 %2 to i16
  store i16 %conv145, ptr %data, align 2, !tbaa !4
  %indvars.iv.next79 = add i64 %indvars.iv78, 1
  %exitcond.not = icmp eq i64 %indvars.iv.next79, 8
  br i1 %exitcond.not, label %for.end, label %for.body

for.end:                                          ; preds = %for.body
  ret void
}

; Function Attrs: nocallback nofree nosync nounwind speculatable willreturn memory(none)
declare i64 @llvm.lrint.i64.f32(float) #0

attributes #0 = { nocallback nofree nosync nounwind speculatable willreturn memory(none) }

!0 = !{!1, !1, i64 0}
!1 = !{!"float", !2, i64 0}
!2 = !{!"omnipotent char", !3, i64 0}
!3 = !{!"Simple C/C++ TBAA"}
!4 = !{!5, !5, i64 0}
!5 = !{!"short", !2, i64 0}
```

artagnon committed:
VPlan: implement VPlan-level constant-folding

Introduce VPlanConstantFolder, a variation of ConstantFolder for VPlan,
and use it in VPBuilder to constant-fold when all the underlying IR
values passed into the API are constants.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1073,13 +1140,14 @@ static void simplifyRecipe(VPRecipeBase &R, VPTypeAnalysis &TypeInfo) {
   for (VPBasicBlock *VPBB : VPBlockUtils::blocksOnly<VPBasicBlock>(RPOT)) {
     for (VPRecipeBase &R : make_early_inc_range(*VPBB)) {
-      simplifyRecipe(R, TypeInfo);
+      simplifyRecipe(R, TypeInfo, DL);
```
Would it be possible to retrieve DL from `Plan`, by using the scalar header IR BB, which is always available, something like `Plan.getScalarHeader()->getIRBasicBlock()->getDataLayout()`?

fhahn reviewed (commented):
Thanks for the latest updates!

fhahn commented on llvm/test/Transforms/LoopVectorize/X86/drop-poison-generating-flags.ll:
```diff
@@ -563,29 +563,24 @@ define void @pr70590_recipe_without_underlying_instr(i64 %n, ptr noalias %dst) {
-; CHECK-NEXT:    [[TMP4:%.*]] = srem i64 3, 0
 ; CHECK-NEXT:    br label [[PRED_SREM_CONTINUE]]
 ; CHECK:       pred.srem.continue:
-; CHECK-NEXT:    [[TMP5:%.*]] = phi i64 [ poison, %vector.body ], [ [[TMP4]], [[PRED_SREM_IF]] ]
```
Can we have a variant of the test where we don't just use the first lane of the simplified replicate recipe, e.g. storing the result outside the replicate region?

Would also be good to have a case where we don't fold to poison, but I am not sure if that's possible.

In that case, could we constant-fold `VPPredPhi poison -> poison`?

artagnon replied:
I've added the additional test, but constant-folding PredPHI seems to be out of scope since it doesn't have an opcode? Could address in a follow-up.

fhahn replied:
Yep, that would need to extra code to handle `PredPhi poison -> poison`. If that would be enough to remove the need to change `VPPredInstPHIRecipe::execute(` and. `VPTransformState::` that would be desireable.

artagnon replied:
That's a good idea, thanks! I've handled PredPHI constant -> constant, and cleaned up the changes in VPlan(Recipes).

lukel97 commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -937,10 +938,77 @@ static void recursivelyDeleteDeadRecipes(VPValue *V) {
   using namespace llvm::VPlanPatternMatch;
 
+  // Constant folding.
+  if (TypeSwitch<VPRecipeBase *, bool>(&R)
+          .Case<VPInstruction, VPWidenRecipe, VPWidenCastRecipe,
+                VPReplicateRecipe>([&](auto *I) {
```
Nit, is this more easily expressed as a ` if (isa<VPInstruction, VPWidenRecipe, VPWidenCastRecipe, VPReplicateRecipe>(R)) {`?

artagnon replied:
We actually need to cast to one of these types for operands() and getOpcode() to be available, so TypeSwitch is necessary.

lukel97 replied:
I thought operands() was a member of VPRecipeBase but I see getOpcode() now, that makes sense

artagnon commented:
Gentle ping.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -938,10 +939,87 @@ static void recursivelyDeleteDeadRecipes(VPValue *V) {
   }
 }
 
+/// Try to fold \p R using TargetFolder to a constant. Will succeed for a
+/// handled \p Opcode if all \p Operands are constant.
```
```suggestion
/// Try to fold \p R using TargetFolder to a constant. Will succeed and return true if \p Opcode is supported and all \p Operands are constant.
```

fhahn reviewed (approved):
LGTM, thanks!

Might be worth adjusting the commit title/message to clarify that this adds constant folding to simplifyRecipe.

artagnon committed:
[VPlanTransforms] Tweak comment
